### PR TITLE
storage: Fix flow controller pending compaction bytes always be zero

### LINKED ISSUE #14392: Flow controller doesn't throttle when compaction pending bytes is high
### What version of TiKV are you using?

master

### What operating system and CPU are you using?

Not related

### Steps to reproduce

Run a so heavy write workload that compaction pending bytes is kept accumulating

### What did you expect?
trigger throttle 

### What did happened?
doesn't trigger throttle

As you can see, the compaction pending bytes is high but the score is always 0
img_v2_36ba7992-4465-44ab-a163-a7c0fb51895g

### PR DESCRIPTION
**Pull request by Connor1996**
### What is changed and how it works?

Issue Number: Close #14392
What's Changed:

```commit-message
Fix the issue that flow control may not work when pending compaction bytes is high.

If the pending compaction bytes is 0, then 0.log2() is -INF which would cause the later
average always be zero even if the pending compaction bytes is already high.
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

### Release note 

```release-note
Fix the issue that flow control may not work when pending compaction bytes is high
```

### CODE REVIEW

**Comment by BusyJay** [comment]
This PR may probably have regression on heavy load scenario.

**Review by zyguan — APPROVED** [review]

File: src/storage/txn/flow_controller/singleton_flow_controller.rs [hunk 1]
```diff
@@ -728,13 +728,16 @@ impl<E: FlowControlFactorStore + Send + 'static> FlowChecker<E> {
...
         // logarithm of pending compaction bytes to make the values fall into
         // a relative small range
-        let num = (self.engine.pending_compaction_bytes(self.region_id, &cf) as f64).log2();
+        let mut num = (self.engine.pending_compaction_bytes(self.region_id, &cf) as f64).log2();
+        if !num.is_finite() {
+            // 0.log2() == -inf, which is not expected
+            num = 0.0;
+        }
         let checker = self.cf_checkers.get_mut(&cf).unwrap();
         checker.long_term_pending_bytes.observe(num);
```
Comment (zyguan) [inline src/storage/txn/flow_controller/singleton_flow_controller.rs:737]: BTW, when `num` is 0, should we treat it as an invalid value and do not count it in `long_term_pending_bytes`.

**Comment by Connor1996** [comment]
> This PR may probably have regression on heavy load scenario.

Yes, though it's an illusion before with a compromise on space amplification

File: src/storage/txn/flow_controller/singleton_flow_controller.rs (hunk 1 above)
Comment (Connor1996) [reply src/storage/txn/flow_controller/singleton_flow_controller.rs:737]: I think it's okay. 0 is valid representing no need to compact, should count in `long_term_pending_bytes`

**Pull request merged**